### PR TITLE
First pass of custom converter

### PR DESCRIPTION
**Pull request by jskeet**
I'm not 100% sure about the null prohibition here. Thoughts welcome on that.

I'll write documentation in a follow-up PR - I expect there may be a bit of back and forth in this one.

### CODE REVIEW

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/SerializationTestData.cs [hunk 1]
```diff
@@ -128,19 +128,28 @@ internal static class SerializationTestData
...
 
-            // Map values (that can be deserialized again): dictionaries, attributed types, expandos (which are just dictionaries)
+            // Map values (that can be deserialized again): dictionaries, attributed types, expandos (which are just dictionaries), custom serialized map-like values
+
+            // Dictionaries
             { new Dictionary<string, object> { { "name", "Jon" }, { "score", 10L } },
                 new Value { MapValue = new MapValue { Fields = { { "name", new Value { StringValue = "Jon" } }, { "score", new Value { IntegerValue = 10L } } } } } },
             { new Dictionary<string, int> { { "A", 10 }, { "B", 20 } },
                 new Value { MapValue = new MapValue { Fields = { { "A", new Value { IntegerValue = 10L } }, { "B", new Value { IntegerValue = 20L } } } } } },
+            // Attributed type (each property has an attributed)
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/SerializationTestData.cs]: nit: has an attribute (without the d in the end)

Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/SerializationTestData.cs]: Will fix, thanks.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Converters/AttributedTypeConverter.cs [hunk 2]
```diff
@@ -83,10 +84,19 @@ private AttributedTypeConverter(BclType targetType) : base(targetType)
...
-        /// (Currently just a constructor call, but this allows for custom conversions in the future.)
         /// </summary>
-        internal static IFirestoreInternalConverter ForType(BclType targetType) => new AttributedTypeConverter(targetType);
-        
+        internal static IFirestoreInternalConverter ForType(BclType targetType)
+        {
+            var typeInfo = targetType.GetTypeInfo();
+            var attribute = typeInfo.GetCustomAttribute<FirestoreDataAttribute>(inherit: false);
+            // This would be an internal library bug. We shouldn't be calling it in this case.
+            GaxPreconditions.CheckState(attribute != null, "Type {0} is not decorated with {1}.", targetType.FullName, nameof(FirestoreDataAttribute));
```
[the remark this answers is not in the document]
Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Converters/AttributedTypeConverter.cs]: Will remove it.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FirestoreDataAttribute.cs [hunk 3]
```diff
@@ -28,6 +28,11 @@ public sealed class FirestoreDataAttribute : Attribute
         /// </summary>
         public UnknownPropertyHandling UnknownPropertyHandling { get; set; }
 
+        /// <summary>
+        /// A custom converter to use for serializing and deserializing the attributed type.
+        /// </summary>
+        public System.Type Converter { get; set; }
```
[the remark this answers is not in the document]
Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FirestoreDataAttribute.cs]: Will do. Wasn't sure what the idiom was here, but AsyncMethodBuilderAttribute uses BuilderType, so we'll follow that.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FirestorePropertyAttribute.cs [hunk 4]
```diff
@@ -28,6 +28,11 @@ public sealed class FirestorePropertyAttribute : Attribute
         /// </summary>
         public string Name { get; set; }
 
+        /// <summary>
+        /// A custom converter to use for this property.
+        /// </summary>
+        public System.Type Converter { get; set; }
```
[the remark this answers is not in the document]
Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FirestorePropertyAttribute.cs]: Yup.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/IFirestoreConverter.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Google.Cloud.Firestore
+{
+    /// <summary>
+    /// Converter used during serialization
```
[the remark this answers is not in the document]
Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/IFirestoreConverter.cs]: More broadly, this needs a lot more documentation :) Will add that.

**Comment by jskeet** [comment]
Added a commit to address issues spotted in review - PTAL, and then we can discuss the null handling more.

**Pull request merged**